### PR TITLE
Add Prometheus metrics relabeling

### PR DESCRIPTION
**Pull request by jangrewe**
# What does this implement/fix?

This implements relabeling of Prometheus metric names and/or IDs using a config block.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2238

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
prometheus:
  relabel:
    device_uptime:
      id: uptime
      name: Uptime
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by glmnet — CHANGES_REQUESTED** [review]
I'm trying to test this and I'm getting a compilation issue, not sure if it's my setup as I'm not doing much core ESPHome lately. 

I just found out `prometheus:` is not used in any **testN.yaml** file
In any case add this component to any of the **testN.yaml** where you see fit,  so the CI builds it, if it goes well I'll check again.

**Review by glmnet — CHANGES_REQUESTED** [review]
short  answer I don't like it, the keys being either ids or names is not nice, also ids might not be what you expect (didn't test but not need to as ids c++ side are generated from name which might be different from an explicit set id in yaml) so I come up with a proposal for you here: https://github.com/esphome/esphome/compare/5edc85b750d81fead2f88a2f859a7801f33617d1...glmnet:esphome:use-cv-useid-as-key
syntax would be:
```yaml
prometheus:
  customize:
    entity_id: # <- validated by esphome validation as the key is a cv.use_id
      id: optional_new_id # not sure if this should have a better validator than string_strict
      name: optional new name
```

I think this is a clean way that can be used in other stuff like web_server to have more per entity data be used in the c++ side

Another way could be what is happening with mqtt but that is trickier as property names could crash 
P.S. I leave c++ implementation to you

**Review by glmnet — APPROVED** [review]
LGTM

File: esphome/components/prometheus/__init__.py [hunk 1]
```diff
@@ -2,23 +2,39 @@
...
 CONFIG_SCHEMA = cv.Schema(
     {
         cv.GenerateID(): cv.declare_id(PrometheusHandler),
         cv.GenerateID(CONF_WEB_SERVER_BASE_ID): cv.use_id(
             web_server_base.WebServerBase
         ),
         cv.Optional(CONF_INCLUDE_INTERNAL, default=False): cv.boolean,
+        cv.Optional(CONF_RELABEL, default={}): cv.Schema(
+            {
+                cv.use_id(EntityBase): CUSTOMIZED_ENTITY,
```
Comment (jesserockz) [inline esphome/components/prometheus/__init__.py:35]: Not arguing that it doesnt work, but might be better to make this a `list` with a `cv.Required("entity"): cv.use_id(EntityBase)`. Having the id as a yaml key seems strange.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**